src/java.base/share/classes/java/util/ImmutableCollections.java

```diff
@@ -1405,6 +1544,130 @@ private Object writeReplace() {
             return new CollSer(CollSer.IMM_MAP, array);
         }
     }
+
+    static final class StableMap<K, V>
+            extends AbstractImmutableMap<K, V> {
+
+        @Stable
+        private final Function<? super K, ? extends V> mapper;
+        @Stable
+        private final Map<K, StableValueImpl<V>> delegate;
+
+        StableMap(Set<K> keys, Function<? super K, ? extends V> mapper) {
+            this.mapper = mapper;
+            this.delegate = StableUtil.map(keys);
+        }
+
+        @Override public boolean              containsKey(Object o) { return delegate.containsKey(o); }
+        @Override public int                  size() { return delegate.size(); }
+        @Override public Set<Map.Entry<K, V>> entrySet() { return new StableMapEntrySet(); }
+
+        @ForceInline
+        @Override
+        public V get(Object key) {
+            return getOrDefault(key, null);
+        }
+
+        @ForceInline
+        @Override
+        public V getOrDefault(Object key, V defaultValue) {
+            final StableValueImpl<V> stable = delegate.get(key);
+            if (stable == null) {
+                return defaultValue;
+            }
+            @SuppressWarnings("unchecked")
+            final K k = (K) key;
+            return stable.orElseSet(new Supplier<V>() {
+                @Override public V get() { return mapper.apply(k); }});
+        }
+
+        @jdk.internal.ValueBased
+        final class StableMapEntrySet extends AbstractImmutableSet<Map.Entry<K, V>> {
```

Comment:
One concern about the (non-static) inner classes is that we can't stable-annotate the immediately enclosing instance - we might need general trusting for java.util, but we need to ensure users aren't hacking collection final fields first.